### ISSUE TITLE
Helidon JPA - GraalVM Native - env variables not read  (Issue 1/2)

### ISSUE THREAD

**Issue by MarcGueury** [helidon-io/helidon#5804]
## Environment Details
* Helidon Version: 3.1
* Helidon MP
* JDK version:17.05
* OS: Linux
* Docker version (if applicable): N/A
  
----------

## Problem Description
If Helidon + JPA + Hibernate + Jackson + Oracle DB driver is used with GraalVM Native 
Passing env variables to modify the microprofile-config.properties does not work.

The goal is to set the connection parameters at runtime for sec purpose.
The problem is probably a lot more generic than with JPA. But I reproduce it repetitively with it.

## Steps to reproduce
Just use the sample generated by Helidon starter.
You need a running database

In your microprofile-config.properties, you can leave it empty if you want too.
```
-------
# Microprofile server properties 
server.port=8080
server.host=0.0.0.0

# Change the following to true to enable the optional MicroProfile Metrics REST.request metrics
metrics.rest-request.enabled=false

# Application properties. This is the default greeting
app.greeting=Hello

# Datasource properties
javax.sql.DataSource.ds1.dataSourceClassName=oracle.jdbc.pool.OracleDataSource
javax.sql.DataSource.ds1.dataSource.url=##JDBC_URL##(invalid value on purpose)
javax.sql.DataSource.ds1.dataSource.user=##DB_USER##
javax.sql.DataSource.ds1.dataSource.password=##DB_PASSWORD##
--------
```
test 1: with JIT (OK):

```
mvn package -DskipTests
cd target
export JAVAX_SQL_DATASOURCE_DS1_DATASOURCE_URL="(jdbc:oracle:thin:@(description=(retry_count=20)(retry_delay=3)(address=(protocol=tcps)(port=1521)(host=xxxxx.adb.eu-frankfurt-1.oraclecloud.com))(connect_data=(service_name=xxxxxxx.adb.oraclecloud.com))(security=(ssl_server_dn_match=yes))))"
export JAVAX_SQL_DATASOURCE_DS1_DATASOURCE_USER=admin
export JAVAX_SQL_DATASOURCE_DS1_DATASOURCE_PASSWORD=xxxxxx
java -jar helidon.jar
-> OK
```
test 2: with Native (BUG):

```
mvn package -Pnative-image -Dnative.image.buildStatic -DskipTests
cd target
export JAVAX_SQL_DATASOURCE_DS1_DATASOURCE_URL="(jdbc:oracle:thin:@(description=(retry_count=20)(retry_delay=3)(address=(protocol=tcps)(port=1521)(host=xxxxx.adb.eu-frankfurt-1.oraclecloud.com))(connect_data=(service_name=xxxxxxx.adb.oraclecloud.com))(security=(ssl_server_dn_match=yes))))"
export JAVAX_SQL_DATASOURCE_DS1_DATASOURCE_USER=admin
export JAVAX_SQL_DATASOURCE_DS1_DATASOURCE_PASSWORD=xxxxxx
./helidon
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.posix.thread.PosixPlatformThreads.pthreadStartRoutine(PosixPlatformThreads.java:203)
Caused by: java.sql.SQLException: Invalid Oracle URL specified
	at oracle.jdbc.datasource.impl.OracleDataSource.getPhysicalConnection(OracleDataSource.java:566)
	at oracle.jdbc.datasource.impl.OracleDataSource.getConnection(OracleDataSource.java:347)
	at oracle.jdbc.datasource.impl.OracleDataSource.getConnectionInternal(OracleDataSource.java:1996)
	at oracle.jdbc.datasource.impl.OracleDataSource.getConnection(OracleDataSource.java:322)
	at oracle.jdbc.datasource.impl.OracleDataSource.getConnection(OracleDataSource.java:283)
	at com.zaxxer.hikari.pool.PoolBase.newConnection(PoolBase.java:359)
	at com.zaxxer.hikari.pool.PoolBase.newPoolEntry(PoolBase.java:201)
	at com.zaxxer.hikari.pool.HikariPool.createPoolEntry(HikariPool.java:470)
	at com.zaxxer.hikari.pool.HikariPool.checkFailFast(HikariPool.java:561)
	... 72 more
```

test 3:
Hardcode the values in microprofile-config.properties
Recompile with GraalVM Native.
It works....

**Comment by ljnelson** [comment]
To be clear: this doesn't concern JPA at all, right? It just concerns configuration and data sources?

**Comment by MarcGueury** [comment]
Yes, I think so too. There is no relation between this behavior and JPA.

**Comment by ljnelson** [comment]
So the data source integration code currently reads configuration to know what data sources there are, so it can figure out what injection points to satisfy in your CDI-based application.  When you build a native image, the configuration is read at build time, because the mechanism used to create data source implementations to inject needs to be done at build time.  At build time, in your example above, you do not supply an environment variable, so `META-INF/microprofile-config.properties` is not overridden (as designed).

If you run:
```
JAVAX_SQL_DATASOURCE_DS1_DATASOURCE_URL="(jdbc:oracle:thin:@(description=(retry_count=20)(retry_delay=3)(address=(protocol=tcps)(port=1521)(host=xxxxx.adb.eu-frankfurt-1.oraclecloud.com))(connect_data=(service_name=xxxxxxx.adb.oraclecloud.com))(security=(ssl_server_dn_match=yes))))" JAVAX_SQL_DATASOURCE_DS1_DATASOURCE_USER=admin JAVAX_SQL_DATASOURCE_DS1_DATASOURCE_PASSWORD=xxxxxx mvn package -Pnative-image -Dnative.image.buildStatic -DskipTests
```
…I expect you will see that the environment variable value is used.  Do you observe this?

**Comment by MarcGueury** [comment]
Thanks. In my case, I build a docker container and pass the DB connection details as kubernetes secret (= env variables) 

If I understand due to that build-time hardcoding, Using META-INF/microprofile-config.properties is not possible anymore in Native mode to pass connection details at run-time. 

Any idea if there is a way to pass user/password/DB_connection without hardcoding them ? The way with META-INF/microprofile-config.properties is the only one I knew...

Thanks.

**Comment by ljnelson** [comment]
(First, `META-INF/microprofile-config.properties` is one of an infinite number of possible MicroProfile Config-compatible [configuration sources](https://github.com/eclipse/microprofile-config/blob/master/spec/src/main/asciidoc/configsources.asciidoc) and is usually reserved for last-ditch defaults. [Environment variables](https://github.com/eclipse/microprofile-config/blob/master/spec/src/main/asciidoc/configsources.asciidoc#environment-variables-mapping-rules) are probably a better choice of MicroProfile Config-compatible configuration source.)

The short answer to your question is: no, at least as of this writing.

If you want to build using one configuration and run with another, you probably shouldn't use native image anyway in this scenario. As mentioned above, the choice of what data sources exist (regardless of how they are configured) must be done at build time.  Currently the configuration for them also happens at build time in native-image scenarios, because the CDI container is validated before being "frozen", and is "thawed" only at native-image run time (i.e. with the data source integration code the way it is, configuration has already been read).

Changing this so that some configuration items (which ones?) are read at build time, and some (which ones? what's the litmus test?) are applied at run time while preserving the validation phase of the container could be done but won't happen anytime soon. You may feel free to file an issue if you want this to be worked on.

Finally, if you are not writing a stateless serverless function (I'm guessing this is the case since you're connecting to a database, so there's state involved) your performance overall will also probably be better if you eschew native image and just go with regular old Java. [`jlink`](https://docs.oracle.com/en/java/javase/19/docs/specs/man/jlink.html) and [application class data sharing](https://docs.oracle.com/en/java/javase/19/vm/class-data-sharing.html) can be your friends here if for some reason size is at an absolute premium.

**Comment by ljnelson** [comment]
I believe this was fixed with #8771. Closing. If you feel this should be reopened, please feel free to comment.
